### DISCUSSION TITLE
Float is rounded to nearest int value

### DISCUSSION THREAD

**Discussion by masrul** [zeux/pugixml#515] [Q&A] [answered]
Hello, 
My XML file has something like that,

```
     <Bond class1="OW" class2="HW" length="0.09572" forceconst="502080.7"/>
```
In my c++ code, 
```
 float bond_fc = bond_type.attribute("forceconst").as_float();  // just relevant line 
```

I expect **bond_fc** to be 502080.7, but it is 502081. Looks like **strtod** is responsible for that. I wonder how to avoid this unwanted rounding to the nearest int value.

**Comment by zeux** [comment, marked as answer]
The value is not rounded to nearest int value - the tool you're using to print/view the value is truncating it.

```
~/pugixml $ cat test.cpp
#include "pugixml.hpp"
#include <stdio.h>

int main()
{
	pugi::xml_document doc;
	doc.load_string("<Bond class1='OW' class2='HW' length='0.09572' forceconst='502080.7'/>");
	float bond_fc = doc.child("Bond").attribute("forceconst").as_float();
	printf("%f\n", bond_fc);
	double bond_fc_d = doc.child("Bond").attribute("forceconst").as_double();
	printf("%f\n", bond_fc_d);
}
~/pugixml $ g++ test.cpp src/pugixml.cpp -I src && ./a.out
502080.687500
502080.700000
```

Single precision floating point numbers can't represent decimals exactly in general, but also have little decimal precision left for large numbers, so the value above has an error (when represented as floats) of ~0.0125. If that error is significant, you can always use double instead.

**Reply by masrul** [reply]
Thanks @zeux
